<a href="https://colab.research.google.com/github/2004sayantan/Heart-Disease-Predictor/blob/main/Heart_disease_predictor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import files
upload=files.upload()

Saving heart_disease_10000_synthetic.xlsx to heart_disease_10000_synthetic.xlsx


In [ ]:
import pandas as pd
df=pd.read_excel("heart_disease_10000_synthetic.xlsx")
df

,Age,Gender,Cholesterol,Blood Pressure,Heart Rate,Smoking,Alcohol Intake,Exercise Hours,Family History,Diabetes,Obesity,Stress Level,Blood Sugar,Exercise Induced Angina,Chest Pain Type,Heart Disease
0,34,Male,145,164,68,Never,Moderate,5,No,No,No,3,136,No,Asymptomatic,0
1,69,Male,281,140,69,Never,NaN,4,No,No,No,7,141,No,Asymptomatic,0
2,63,Female,283,127,59,Never,Moderate,4,Yes,No,Yes,8,98,No,Non-anginal Pain,0
3,52,Female,209,179,75,Current,Heavy,6,No,No,No,4,157,No,Non-anginal Pain,0
4,52,Male,253,130,72,Never,Heavy,5,No,No,No,1,108,No,Typical Angina,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,54,Male,241,157,73,Former,NaN,10,Yes,No,No,8,148,Yes,Atypical Angina,0
9996,80,Male,262,104,85,Never,Heavy,5,Yes,No,Yes,9,163,No,Atypical Angina,1
9997,48,Male,198,137,61,Former,NaN,7,Yes,No,No,1,157,No,Atypical Angina,0
9998,58,Male,260,80,84,Never,NaN,5,No,No,No,7,65,No,Asymptomatic,0


In [ ]:
df.shape

(10000, 16)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Age                      10000 non-null  int64 
 1   Gender                   10000 non-null  object
 2   Cholesterol              10000 non-null  int64 
 3   Blood Pressure           10000 non-null  int64 
 4   Heart Rate               10000 non-null  int64 
 5   Smoking                  10000 non-null  object
 6   Alcohol Intake           5790 non-null   object
 7   Exercise Hours           10000 non-null  int64 
 8   Family History           10000 non-null  object
 9   Diabetes                 10000 non-null  object
 10  Obesity                  10000 non-null  object
 11  Stress Level             10000 non-null  int64 
 12  Blood Sugar              10000 non-null  int64 
 13  Exercise Induced Angina  10000 non-null  object
 14  Chest Pain Type          10000 non-null

In [ ]:
df.describe()

,Age,Cholesterol,Blood Pressure,Heart Rate,Exercise Hours,Stress Level,Blood Sugar,Heart Disease
count,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000
mean,54.816000,230.803300,130.788500,74.984800,5.019200,5.494200,121.028100,0.189600
std,14.645211,44.557722,26.771594,11.866265,2.420747,2.856953,33.314377,0.392004
min,30.000000,130.000000,80.000000,50.000000,0.000000,1.000000,65.000000,0.000000
25%,42.000000,200.000000,112.000000,67.000000,3.000000,3.000000,97.000000,0.000000
50%,55.000000,230.000000,131.000000,75.000000,5.000000,5.000000,120.000000,0.000000
75%,68.000000,261.000000,149.000000,83.000000,7.000000,8.000000,144.000000,0.000000
max,80.000000,360.000000,190.000000,110.000000,10.000000,10.000000,220.000000,1.000000


# Handle Null values

In [ ]:
df.head(10)

,Age,Gender,Cholesterol,Blood Pressure,Heart Rate,Smoking,Alcohol Intake,Exercise Hours,Family History,Diabetes,Obesity,Stress Level,Blood Sugar,Exercise Induced Angina,Chest Pain Type,Heart Disease
0,34,Male,145,164,68,Never,Moderate,5,No,No,No,3,136,No,Asymptomatic,0
1,69,Male,281,140,69,Never,NaN,4,No,No,No,7,141,No,Asymptomatic,0
2,63,Female,283,127,59,Never,Moderate,4,Yes,No,Yes,8,98,No,Non-anginal Pain,0
3,52,Female,209,179,75,Current,Heavy,6,No,No,No,4,157,No,Non-anginal Pain,0
4,52,Male,253,130,72,Never,Heavy,5,No,No,No,1,108,No,Typical Angina,0
5,73,Female,220,156,97,Current,Heavy,7,No,No,No,4,108,No,Atypical Angina,0
6,34,Female,241,185,85,Former,NaN,2,No,No,No,7,132,Yes,Typical Angina,0
7,65,Male,178,153,69,Never,Moderate,5,No,No,No,6,74,No,Asymptomatic,0
8,40,Female,258,102,55,Current,Heavy,3,No,No,Yes,1,101,No,Atypical Angina,0
9,34,Male,156,95,50,Never,Heavy,5,No,No,No,6,114,No,Typical Angina,0


In [ ]:
alcohol_intake=df.groupby(['Alcohol Intake'])['Alcohol Intake'].count()
alcohol_intake

,Alcohol Intake
Alcohol Intake,
Heavy,2071
Moderate,3719


In [ ]:
df.isnull().sum()

,0
Age,0
Gender,0
Cholesterol,0
Blood Pressure,0
Heart Rate,0
Smoking,0
Alcohol Intake,4210
Exercise Hours,0
Family History,0
Diabetes,0


In [ ]:
df['Alcohol Intake'].unique()

array(['Moderate', nan, 'Heavy'], dtype=object)

In [ ]:
df['Alcohol Intake'].value_counts(dropna=False)

,count
Alcohol Intake,
NaN,4210
Moderate,3719
Heavy,2071


In [ ]:
df['Alcohol Intake'] = df['Alcohol Intake'].fillna('Unknown')
df['Alcohol Intake'].isnull().sum()

np.int64(0)

In [ ]:
alcohol_intake=df.groupby(['Alcohol Intake'])['Alcohol Intake'].count()
alcohol_intake

,Alcohol Intake
Alcohol Intake,
Heavy,2071
Moderate,3719
Unknown,4210


In [ ]:
df['Alcohol Intake'].value_counts()

,count
Alcohol Intake,
Unknown,4210
Moderate,3719
Heavy,2071


In [ ]:
df.head(10)

,Age,Gender,Cholesterol,Blood Pressure,Heart Rate,Smoking,Alcohol Intake,Exercise Hours,Family History,Diabetes,Obesity,Stress Level,Blood Sugar,Exercise Induced Angina,Chest Pain Type,Heart Disease
0,34,Male,145,164,68,Never,Moderate,5,No,No,No,3,136,No,Asymptomatic,0
1,69,Male,281,140,69,Never,Unknown,4,No,No,No,7,141,No,Asymptomatic,0
2,63,Female,283,127,59,Never,Moderate,4,Yes,No,Yes,8,98,No,Non-anginal Pain,0
3,52,Female,209,179,75,Current,Heavy,6,No,No,No,4,157,No,Non-anginal Pain,0
4,52,Male,253,130,72,Never,Heavy,5,No,No,No,1,108,No,Typical Angina,0
5,73,Female,220,156,97,Current,Heavy,7,No,No,No,4,108,No,Atypical Angina,0
6,34,Female,241,185,85,Former,Unknown,2,No,No,No,7,132,Yes,Typical Angina,0
7,65,Male,178,153,69,Never,Moderate,5,No,No,No,6,74,No,Asymptomatic,0
8,40,Female,258,102,55,Current,Heavy,3,No,No,Yes,1,101,No,Atypical Angina,0
9,34,Male,156,95,50,Never,Heavy,5,No,No,No,6,114,No,Typical Angina,0


# Handle duplicate data

In [ ]:
df.duplicated().sum()

np.int64(0)

In [ ]:
df.head()

,Age,Gender,Cholesterol,Blood Pressure,Heart Rate,Smoking,Alcohol Intake,Exercise Hours,Family History,Diabetes,Obesity,Stress Level,Blood Sugar,Exercise Induced Angina,Chest Pain Type,Heart Disease
0,34,Male,145,164,68,Never,Moderate,5,No,No,No,3,136,No,Asymptomatic,0
1,69,Male,281,140,69,Never,Unknown,4,No,No,No,7,141,No,Asymptomatic,0
2,63,Female,283,127,59,Never,Moderate,4,Yes,No,Yes,8,98,No,Non-anginal Pain,0
3,52,Female,209,179,75,Current,Heavy,6,No,No,No,4,157,No,Non-anginal Pain,0
4,52,Male,253,130,72,Never,Heavy,5,No,No,No,1,108,No,Typical Angina,0


In [ ]:
df.columns

Index(['Age', 'Gender', 'Cholesterol', 'Blood Pressure', 'Heart Rate',
       'Smoking', 'Alcohol Intake', 'Exercise Hours', 'Family History',
       'Diabetes', 'Obesity', 'Stress Level', 'Blood Sugar',
       'Exercise Induced Angina', 'Chest Pain Type', 'Heart Disease'],
      dtype='object')

# Encode gender using OneHot Encoding

In [ ]:
gender_data=df.groupby(['Gender'])['Gender'].count()
gender_data

,Gender
Gender,
Female,4989
Male,5011


In [ ]:
from sklearn.preprocessing import OneHotEncoder
import pandas as pd
ohe=OneHotEncoder(sparse_output=False)

on_gender=ohe.fit_transform(df[['Gender']])
encoded_gender=pd.DataFrame(
    on_gender,
    columns=ohe.get_feature_names_out(['Gender']),
    index=df.index
)
df=pd.concat([df,encoded_gender],axis=1)
df


,Age,Gender,Cholesterol,Blood Pressure,Heart Rate,Smoking,Alcohol Intake,Exercise Hours,Family History,Diabetes,Obesity,Stress Level,Blood Sugar,Exercise Induced Angina,Chest Pain Type,Heart Disease,Gender_Female,Gender_Male
0,34,Male,145,164,68,Never,Moderate,5,No,No,No,3,136,No,Asymptomatic,0,0.0,1.0
1,69,Male,281,140,69,Never,Unknown,4,No,No,No,7,141,No,Asymptomatic,0,0.0,1.0
2,63,Female,283,127,59,Never,Moderate,4,Yes,No,Yes,8,98,No,Non-anginal Pain,0,1.0,0.0
3,52,Female,209,179,75,Current,Heavy,6,No,No,No,4,157,No,Non-anginal Pain,0,1.0,0.0
4,52,Male,253,130,72,Never,Heavy,5,No,No,No,1,108,No,Typical Angina,0,0.0,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,54,Male,241,157,73,Former,Unknown,10,Yes,No,No,8,148,Yes,Atypical Angina,0,0.0,1.0
9996,80,Male,262,104,85,Never,Heavy,5,Yes,No,Yes,9,163,No,Atypical Angina,1,0.0,1.0
9997,48,Male,198,137,61,Former,Unknown,7,Yes,No,No,1,157,No,Atypical Angina,0,0.0,1.0
9998,58,Male,260,80,84,Never,Unknown,5,No,No,No,7,65,No,Asymptomatic,0,0.0,1.0


# Encode Smoking using OneHotEncoding

In [ ]:
smoking_data=df.groupby(['Smoking'])['Smoking'].count()
smoking_data

,Smoking
Smoking,
Current,2801
Former,2215
Never,4984


In [ ]:
from operator import index
from sklearn.preprocessing import OneHotEncoder
import pandas as pd

ohe=OneHotEncoder(sparse_output=False)
on_smoking=ohe.fit_transform(df[['Smoking']])
encoded_smoking=pd.DataFrame(
    on_smoking,
    columns=ohe.get_feature_names_out(['Smoking']),
    index=df.index
)
df=pd.concat([df,encoded_smoking],axis=1)
df

,Age,Gender,Cholesterol,Blood Pressure,Heart Rate,Smoking,Alcohol Intake,Exercise Hours,Family History,Diabetes,...,Stress Level,Blood Sugar,Exercise Induced Angina,Chest Pain Type,Heart Disease,Gender_Female,Gender_Male,Smoking_Current,Smoking_Former,Smoking_Never
0,34,Male,145,164,68,Never,Moderate,5,No,No,...,3,136,No,Asymptomatic,0,0.0,1.0,0.0,0.0,1.0
1,69,Male,281,140,69,Never,Unknown,4,No,No,...,7,141,No,Asymptomatic,0,0.0,1.0,0.0,0.0,1.0
2,63,Female,283,127,59,Never,Moderate,4,Yes,No,...,8,98,No,Non-anginal Pain,0,1.0,0.0,0.0,0.0,1.0
3,52,Female,209,179,75,Current,Heavy,6,No,No,...,4,157,No,Non-anginal Pain,0,1.0,0.0,1.0,0.0,0.0
4,52,Male,253,130,72,Never,Heavy,5,No,No,...,1,108,No,Typical Angina,0,0.0,1.0,0.0,0.0,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,54,Male,241,157,73,Former,Unknown,10,Yes,No,...,8,148,Yes,Atypical Angina,0,0.0,1.0,0.0,1.0,0.0
9996,80,Male,262,104,85,Never,Heavy,5,Yes,No,...,9,163,No,Atypical Angina,1,0.0,1.0,0.0,0.0,1.0
9997,48,Male,198,137,61,Former,Unknown,7,Yes,No,...,1,157,No,Atypical Angina,0,0.0,1.0,0.0,1.0,0.0
9998,58,Male,260,80,84,Never,Unknown,5,No,No,...,7,65,No,Asymptomatic,0,0.0,1.0,0.0,0.0,1.0


# Encode Alcohol Intake using OneHot Encoding

In [ ]:
alcohol_intake_data=df.groupby(['Alcohol Intake'])['Alcohol Intake'].count()
alcohol_intake_data

,Alcohol Intake
Alcohol Intake,
Heavy,2071
Moderate,3719
Unknown,4210


In [ ]:
from sklearn.preprocessing import OneHotEncoder
import pandas as pd

ohe=OneHotEncoder(sparse_output=False)
on_alcohol_intake=ohe.fit_transform(df[['Alcohol Intake']])
encoded_alchol_intake=pd.DataFrame(
    on_alcohol_intake,
    columns=ohe.get_feature_names_out(['Alcohol Intake']),
    index=df.index
)
df=pd.concat([df,encoded_alchol_intake],axis=1)
df

,Age,Gender,Cholesterol,Blood Pressure,Heart Rate,Smoking,Alcohol Intake,Exercise Hours,Family History,Diabetes,...,Chest Pain Type,Heart Disease,Gender_Female,Gender_Male,Smoking_Current,Smoking_Former,Smoking_Never,Alcohol Intake_Heavy,Alcohol Intake_Moderate,Alcohol Intake_Unknown
0,34,Male,145,164,68,Never,Moderate,5,No,No,...,Asymptomatic,0,0.0,1.0,0.0,0.0,1.0,0.0,1.0,0.0
1,69,Male,281,140,69,Never,Unknown,4,No,No,...,Asymptomatic,0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0
2,63,Female,283,127,59,Never,Moderate,4,Yes,No,...,Non-anginal Pain,0,1.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0
3,52,Female,209,179,75,Current,Heavy,6,No,No,...,Non-anginal Pain,0,1.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0
4,52,Male,253,130,72,Never,Heavy,5,No,No,...,Typical Angina,0,0.0,1.0,0.0,0.0,1.0,1.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,54,Male,241,157,73,Former,Unknown,10,Yes,No,...,Atypical Angina,0,0.0,1.0,0.0,1.0,0.0,0.0,0.0,1.0
9996,80,Male,262,104,85,Never,Heavy,5,Yes,No,...,Atypical Angina,1,0.0,1.0,0.0,0.0,1.0,1.0,0.0,0.0
9997,48,Male,198,137,61,Former,Unknown,7,Yes,No,...,Atypical Angina,0,0.0,1.0,0.0,1.0,0.0,0.0,0.0,1.0
9998,58,Male,260,80,84,Never,Unknown,5,No,No,...,Asymptomatic,0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0


# Encode Family History using OneHot Encoding

In [ ]:
family_history_data=df.groupby(['Family History'])['Family History'].count()
family_history_data

,Family History
Family History,
No,6573
Yes,3427


In [ ]:
from sklearn.preprocessing import OneHotEncoder
import pandas as pd

ohe=OneHotEncoder(sparse_output=False)

on_family_hist=ohe.fit_transform(df[['Family History']])
encoded_family_hist=pd.DataFrame(
    on_family_hist,
    columns=ohe.get_feature_names_out(['Family History']),
    index=df.index
)
df=pd.concat([df,encoded_family_hist],axis=1)
df

,Age,Gender,Cholesterol,Blood Pressure,Heart Rate,Smoking,Alcohol Intake,Exercise Hours,Family History,Diabetes,...,Gender_Female,Gender_Male,Smoking_Current,Smoking_Former,Smoking_Never,Alcohol Intake_Heavy,Alcohol Intake_Moderate,Alcohol Intake_Unknown,Family History_No,Family History_Yes
0,34,Male,145,164,68,Never,Moderate,5,No,No,...,0.0,1.0,0.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
1,69,Male,281,140,69,Never,Unknown,4,No,No,...,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0
2,63,Female,283,127,59,Never,Moderate,4,Yes,No,...,1.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0
3,52,Female,209,179,75,Current,Heavy,6,No,No,...,1.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0
4,52,Male,253,130,72,Never,Heavy,5,No,No,...,0.0,1.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,54,Male,241,157,73,Former,Unknown,10,Yes,No,...,0.0,1.0,0.0,1.0,0.0,0.0,0.0,1.0,0.0,1.0
9996,80,Male,262,104,85,Never,Heavy,5,Yes,No,...,0.0,1.0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,1.0
9997,48,Male,198,137,61,Former,Unknown,7,Yes,No,...,0.0,1.0,0.0,1.0,0.0,0.0,0.0,1.0,0.0,1.0
9998,58,Male,260,80,84,Never,Unknown,5,No,No,...,0.0,1.0,0.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0


# Encode Diabetes using OneHot Encoding

In [ ]:
diabetes_data=df.groupby(['Diabetes'])['Diabetes'].count()
diabetes_data

,Diabetes
Diabetes,
No,7989
Yes,2011


In [ ]:
from sklearn.preprocessing import OneHotEncoder
import pandas as pd

ohe=OneHotEncoder(sparse_output=False)

on_diabetes=ohe.fit_transform(df[['Diabetes']])
encoded_diabetes=pd.DataFrame(
    on_diabetes,
    columns=ohe.get_feature_names_out(['Diabetes']),
    index=df.index
)
df=pd.concat([df,encoded_diabetes],axis=1)
df

,Age,Gender,Cholesterol,Blood Pressure,Heart Rate,Smoking,Alcohol Intake,Exercise Hours,Family History,Diabetes,...,Smoking_Current,Smoking_Former,Smoking_Never,Alcohol Intake_Heavy,Alcohol Intake_Moderate,Alcohol Intake_Unknown,Family History_No,Family History_Yes,Diabetes_No,Diabetes_Yes
0,34,Male,145,164,68,Never,Moderate,5,No,No,...,0.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
1,69,Male,281,140,69,Never,Unknown,4,No,No,...,0.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0
2,63,Female,283,127,59,Never,Moderate,4,Yes,No,...,0.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0
3,52,Female,209,179,75,Current,Heavy,6,No,No,...,1.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,1.0,0.0
4,52,Male,253,130,72,Never,Heavy,5,No,No,...,0.0,0.0,1.0,1.0,0.0,0.0,1.0,0.0,1.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,54,Male,241,157,73,Former,Unknown,10,Yes,No,...,0.0,1.0,0.0,0.0,0.0,1.0,0.0,1.0,1.0,0.0
9996,80,Male,262,104,85,Never,Heavy,5,Yes,No,...,0.0,0.0,1.0,1.0,0.0,0.0,0.0,1.0,1.0,0.0
9997,48,Male,198,137,61,Former,Unknown,7,Yes,No,...,0.0,1.0,0.0,0.0,0.0,1.0,0.0,1.0,1.0,0.0
9998,58,Male,260,80,84,Never,Unknown,5,No,No,...,0.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0


In [ ]:
df.columns

Index(['Age', 'Gender', 'Cholesterol', 'Blood Pressure', 'Heart Rate',
       'Smoking', 'Alcohol Intake', 'Exercise Hours', 'Family History',
       'Diabetes', 'Obesity', 'Stress Level', 'Blood Sugar',
       'Exercise Induced Angina', 'Chest Pain Type', 'Heart Disease',
       'Gender_Female', 'Gender_Male', 'Smoking_Current', 'Smoking_Former',
       'Smoking_Never', 'Alcohol Intake_Heavy', 'Alcohol Intake_Moderate',
       'Alcohol Intake_Unknown', 'Family History_No', 'Family History_Yes',
       'Diabetes_No', 'Diabetes_Yes'],
      dtype='object')

In [ ]:
df[['Age', 'Gender', 'Cholesterol', 'Blood Pressure', 'Heart Rate',
       'Smoking', 'Alcohol Intake', 'Exercise Hours', 'Family History',
       'Diabetes', 'Obesity', 'Stress Level', 'Blood Sugar',
       'Exercise Induced Angina', 'Chest Pain Type', 'Heart Disease']]

,Age,Gender,Cholesterol,Blood Pressure,Heart Rate,Smoking,Alcohol Intake,Exercise Hours,Family History,Diabetes,Obesity,Stress Level,Blood Sugar,Exercise Induced Angina,Chest Pain Type,Heart Disease
0,34,Male,145,164,68,Never,Moderate,5,No,No,No,3,136,No,Asymptomatic,0
1,69,Male,281,140,69,Never,Unknown,4,No,No,No,7,141,No,Asymptomatic,0
2,63,Female,283,127,59,Never,Moderate,4,Yes,No,Yes,8,98,No,Non-anginal Pain,0
3,52,Female,209,179,75,Current,Heavy,6,No,No,No,4,157,No,Non-anginal Pain,0
4,52,Male,253,130,72,Never,Heavy,5,No,No,No,1,108,No,Typical Angina,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,54,Male,241,157,73,Former,Unknown,10,Yes,No,No,8,148,Yes,Atypical Angina,0
9996,80,Male,262,104,85,Never,Heavy,5,Yes,No,Yes,9,163,No,Atypical Angina,1
9997,48,Male,198,137,61,Former,Unknown,7,Yes,No,No,1,157,No,Atypical Angina,0
9998,58,Male,260,80,84,Never,Unknown,5,No,No,No,7,65,No,Asymptomatic,0


# Encoding Obesity using OneHot Encoding

In [ ]:
obesity_data=df.groupby(['Obesity'])['Obesity'].count()
obesity_data

,Obesity
Obesity,
No,6528
Yes,3472


In [ ]:
from sklearn.preprocessing import OneHotEncoder
import pandas as pd

ohe=OneHotEncoder(sparse_output=False)

on_Obesity=ohe.fit_transform(df[['Obesity']])
encoded_Obesity=pd.DataFrame(
    on_Obesity,
    columns=ohe.get_feature_names_out(['Obesity']),
    index=df.index
)
df=pd.concat([df,encoded_Obesity],axis=1)
df

,Age,Gender,Cholesterol,Blood Pressure,Heart Rate,Smoking,Alcohol Intake,Exercise Hours,Family History,Diabetes,...,Smoking_Never,Alcohol Intake_Heavy,Alcohol Intake_Moderate,Alcohol Intake_Unknown,Family History_No,Family History_Yes,Diabetes_No,Diabetes_Yes,Obesity_No,Obesity_Yes
0,34,Male,145,164,68,Never,Moderate,5,No,No,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
1,69,Male,281,140,69,Never,Unknown,4,No,No,...,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0
2,63,Female,283,127,59,Never,Moderate,4,Yes,No,...,1.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0
3,52,Female,209,179,75,Current,Heavy,6,No,No,...,0.0,1.0,0.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
4,52,Male,253,130,72,Never,Heavy,5,No,No,...,1.0,1.0,0.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,54,Male,241,157,73,Former,Unknown,10,Yes,No,...,0.0,0.0,0.0,1.0,0.0,1.0,1.0,0.0,1.0,0.0
9996,80,Male,262,104,85,Never,Heavy,5,Yes,No,...,1.0,1.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0
9997,48,Male,198,137,61,Former,Unknown,7,Yes,No,...,0.0,0.0,0.0,1.0,0.0,1.0,1.0,0.0,1.0,0.0
9998,58,Male,260,80,84,Never,Unknown,5,No,No,...,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0


# Encode Exercise Induced Angina using One Hot encoding

In [ ]:
eia_data=df.groupby(['Exercise Induced Angina'])['Exercise Induced Angina'].count()
eia_data

,Exercise Induced Angina
Exercise Induced Angina,
No,6580
Yes,3420


In [ ]:
from sklearn.preprocessing import OneHotEncoder
import pandas as pd

ohe=OneHotEncoder(sparse_output=False)

on_eia=ohe.fit_transform(df[['Exercise Induced Angina']])
encoded_eia=pd.DataFrame(
    on_eia,
    columns=ohe.get_feature_names_out(['Exercise Induced Angina']),
    index=df.index
)
df=pd.concat([df,encoded_eia],axis=1)
df

,Age,Gender,Cholesterol,Blood Pressure,Heart Rate,Smoking,Alcohol Intake,Exercise Hours,Family History,Diabetes,...,Alcohol Intake_Moderate,Alcohol Intake_Unknown,Family History_No,Family History_Yes,Diabetes_No,Diabetes_Yes,Obesity_No,Obesity_Yes,Exercise Induced Angina_No,Exercise Induced Angina_Yes
0,34,Male,145,164,68,Never,Moderate,5,No,No,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
1,69,Male,281,140,69,Never,Unknown,4,No,No,...,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
2,63,Female,283,127,59,Never,Moderate,4,Yes,No,...,1.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0,1.0,0.0
3,52,Female,209,179,75,Current,Heavy,6,No,No,...,0.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
4,52,Male,253,130,72,Never,Heavy,5,No,No,...,0.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,54,Male,241,157,73,Former,Unknown,10,Yes,No,...,0.0,1.0,0.0,1.0,1.0,0.0,1.0,0.0,0.0,1.0
9996,80,Male,262,104,85,Never,Heavy,5,Yes,No,...,0.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0,1.0,0.0
9997,48,Male,198,137,61,Former,Unknown,7,Yes,No,...,0.0,1.0,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0
9998,58,Male,260,80,84,Never,Unknown,5,No,No,...,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0


# Encode Chest Pain Type using One Hot encoding

In [ ]:
chest_pain_data=df.groupby(['Chest Pain Type'])['Chest Pain Type'].count()
chest_pain_data

,Chest Pain Type
Chest Pain Type,
Asymptomatic,2431
Atypical Angina,2720
Non-anginal Pain,2806
Typical Angina,2043


In [ ]:
from sklearn.preprocessing import OneHotEncoder
import pandas as pd

ohe=OneHotEncoder(sparse_output=False)

on_chest_pain_data=ohe.fit_transform(df[['Chest Pain Type']])
encoded_chest_pain_data=pd.DataFrame(
    on_chest_pain_data,
    columns=ohe.get_feature_names_out(['Chest Pain Type']),
    index=df.index
)
df=pd.concat([df,encoded_chest_pain_data],axis=1)
df

,Age,Gender,Cholesterol,Blood Pressure,Heart Rate,Smoking,Alcohol Intake,Exercise Hours,Family History,Diabetes,...,Diabetes_No,Diabetes_Yes,Obesity_No,Obesity_Yes,Exercise Induced Angina_No,Exercise Induced Angina_Yes,Chest Pain Type_Asymptomatic,Chest Pain Type_Atypical Angina,Chest Pain Type_Non-anginal Pain,Chest Pain Type_Typical Angina
0,34,Male,145,164,68,Never,Moderate,5,No,No,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,0.0
1,69,Male,281,140,69,Never,Unknown,4,No,No,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,0.0
2,63,Female,283,127,59,Never,Moderate,4,Yes,No,...,1.0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,1.0,0.0
3,52,Female,209,179,75,Current,Heavy,6,No,No,...,1.0,0.0,1.0,0.0,1.0,0.0,0.0,0.0,1.0,0.0
4,52,Male,253,130,72,Never,Heavy,5,No,No,...,1.0,0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,54,Male,241,157,73,Former,Unknown,10,Yes,No,...,1.0,0.0,1.0,0.0,0.0,1.0,0.0,1.0,0.0,0.0
9996,80,Male,262,104,85,Never,Heavy,5,Yes,No,...,1.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0,0.0,0.0
9997,48,Male,198,137,61,Former,Unknown,7,Yes,No,...,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0
9998,58,Male,260,80,84,Never,Unknown,5,No,No,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,0.0


In [ ]:
df[['Age', 'Gender', 'Cholesterol', 'Blood Pressure', 'Heart Rate',
       'Smoking', 'Alcohol Intake', 'Exercise Hours', 'Family History',
       'Diabetes', 'Obesity', 'Stress Level', 'Blood Sugar',
       'Exercise Induced Angina', 'Chest Pain Type', 'Heart Disease']]

,Age,Gender,Cholesterol,Blood Pressure,Heart Rate,Smoking,Alcohol Intake,Exercise Hours,Family History,Diabetes,Obesity,Stress Level,Blood Sugar,Exercise Induced Angina,Chest Pain Type,Heart Disease
0,34,Male,145,164,68,Never,Moderate,5,No,No,No,3,136,No,Asymptomatic,0
1,69,Male,281,140,69,Never,Unknown,4,No,No,No,7,141,No,Asymptomatic,0
2,63,Female,283,127,59,Never,Moderate,4,Yes,No,Yes,8,98,No,Non-anginal Pain,0
3,52,Female,209,179,75,Current,Heavy,6,No,No,No,4,157,No,Non-anginal Pain,0
4,52,Male,253,130,72,Never,Heavy,5,No,No,No,1,108,No,Typical Angina,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,54,Male,241,157,73,Former,Unknown,10,Yes,No,No,8,148,Yes,Atypical Angina,0
9996,80,Male,262,104,85,Never,Heavy,5,Yes,No,Yes,9,163,No,Atypical Angina,1
9997,48,Male,198,137,61,Former,Unknown,7,Yes,No,No,1,157,No,Atypical Angina,0
9998,58,Male,260,80,84,Never,Unknown,5,No,No,No,7,65,No,Asymptomatic,0


# Logistic Regression model with OneHotEncoding

In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import OneHotEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import re


# Function to ignore spaces, signs and uppercase/lowercase
def clean_text(text):
    return re.sub(r'[^a-zA-Z0-9]', '', text).lower()


# One Hot Encoding
ohe = OneHotEncoder(sparse_output=False)

encoded = ohe.fit_transform(df[['Gender','Smoking','Alcohol Intake','Family History',
                                'Obesity','Exercise Induced Angina','Chest Pain Type']])

encoded_df = pd.DataFrame(
    encoded,
    columns=ohe.get_feature_names_out(['Gender','Smoking','Alcohol Intake','Family History',
                                        'Obesity','Exercise Induced Angina','Chest Pain Type']),
    index=df.index
)

X=pd.concat([df[['Age','Cholesterol','Blood Pressure','Heart Rate','Exercise Hours','Stress Level','Blood Sugar']],encoded_df],axis=1)
y=df['Heart Disease']

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=90)

model=LogisticRegression()
model.fit(X_train,y_train)
on_logistic_predict=model.predict(X_test)

print(on_logistic_predict)

print("-----------------------------------------")

l_accuracy = accuracy_score(y_test, on_logistic_predict)

print("Accuracy:", l_accuracy)
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, on_logistic_predict))

print("\nClassification Report:")
print(classification_report(y_test, on_logistic_predict))



print("-------------------------------------------")


[0 0 0 ... 0 0 0]
-----------------------------------------
Accuracy: 0.8105

Confusion Matrix:
[[1573   34]
 [ 345   48]]

Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.98      0.89      1607
           1       0.59      0.12      0.20       393

    accuracy                           0.81      2000
   macro avg       0.70      0.55      0.55      2000
weighted avg       0.77      0.81      0.76      2000

-------------------------------------------


/usr/local/lib/python3.13/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [ ]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Features
numeric_cols = [
    'Age',
    'Cholesterol',
    'Blood Pressure',
    'Heart Rate',
    'Exercise Hours',
    'Stress Level',
    'Blood Sugar'
]

categorical_cols = [
    'Gender',
    'Smoking',
    'Alcohol Intake',
    'Family History',
    'Obesity',
    'Exercise Induced Angina',
    'Chest Pain Type'
]

X = df[numeric_cols + categorical_cols]
y = df['Heart Disease']

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

# Preprocessing
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_cols),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
    ]
)

# Model pipeline
model = Pipeline([
    ('preprocessor', preprocessor),
    ('classifier', LogisticRegression(max_iter=1000))
])

# Train
model.fit(X_train, y_train)

# Prediction
y_pred = model.predict(X_test)

# Evaluation
lo_accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", lo_accuracy)

print("-----------------------------------------")

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("-----------------------------------------")

print("Classification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 0.8195
-----------------------------------------
Confusion Matrix:
[[1577   44]
 [ 317   62]]
-----------------------------------------
Classification Report:
              precision    recall  f1-score   support

           0       0.83      0.97      0.90      1621
           1       0.58      0.16      0.26       379

    accuracy                           0.82      2000
   macro avg       0.71      0.57      0.58      2000
weighted avg       0.79      0.82      0.78      2000



0 = No heart disease
1 = Heart disease

Accuracy = 81.95%

1577 — Correctly identified healthy people

1,577 people actually did NOT have heart disease, and the model correctly predicted No Heart Disease.

44 — False alarm

44 people did NOT have heart disease, but the model predicted Heart Disease.

317 — Missed heart disease cases

This is the most important problem.

317 people actually HAD heart disease, but the model predicted No Heart Disease

62 — Correctly detected heart disease

62 people actually had heart disease, and the model correctly predicted Heart Disease.

# Decision Tree  using OneHot Encoding

In [ ]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.tree import DecisionTreeClassifier

# Features
numeric_cols = [
    'Age',
    'Cholesterol',
    'Blood Pressure',
    'Heart Rate',
    'Exercise Hours',
    'Stress Level',
    'Blood Sugar'
]

categorical_cols = [
    'Gender',
    'Smoking',
    'Alcohol Intake',
    'Family History',
    'Obesity',
    'Exercise Induced Angina',
    'Chest Pain Type'
]

X = df[numeric_cols + categorical_cols]
y = df['Heart Disease']

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

# Preprocessing
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_cols),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
    ]
)

# Model pipeline
model = Pipeline([
    ('preprocessor', preprocessor),
    ('classifier', DecisionTreeClassifier(random_state=42))
])

# Train
model.fit(X_train, y_train)

# Prediction
y_pred = model.predict(X_test)

# Evaluation
d_accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", d_accuracy)

print("-----------------------------------------")

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("-----------------------------------------")

print("Classification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 0.713
-----------------------------------------
Confusion Matrix:
[[1292  329]
 [ 245  134]]
-----------------------------------------
Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.80      0.82      1621
           1       0.29      0.35      0.32       379

    accuracy                           0.71      2000
   macro avg       0.57      0.58      0.57      2000
weighted avg       0.74      0.71      0.72      2000



0 = No Heart Disease
1 = Heart Disease

Accuracy = 71.3%

1292 — Correctly identified healthy people

1,292 people did NOT have heart disease, and your model correctly predicted No Heart Disease.

329 — False alarms

329 people did NOT have heart disease, but your model predicted Heart Disease.

These are false positives.

245 — Missed heart disease cases

245 people actually HAD heart disease, but the model predicted No Heart Disease.

These are false negatives.

134 — Correctly detected heart disease

134 people actually had heart disease, and the model correctly predicted Heart Disease.

That's much better than your previous model, which correctly detected only 62.

# KNeighborsClassifier using OneHot Encoding

In [ ]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.neighbors import KNeighborsClassifier

# Features
numeric_cols = [
    'Age',
    'Cholesterol',
    'Blood Pressure',
    'Heart Rate',
    'Exercise Hours',
    'Stress Level',
    'Blood Sugar'
]

categorical_cols = [
    'Gender',
    'Smoking',
    'Alcohol Intake',
    'Family History',
    'Obesity',
    'Exercise Induced Angina',
    'Chest Pain Type'
]

X = df[numeric_cols + categorical_cols]
y = df['Heart Disease']

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

# Preprocessing
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_cols),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
    ]
)

# Model pipeline
model = Pipeline([
    ('preprocessor', preprocessor),
    ('classifier', KNeighborsClassifier(n_neighbors=5))
])

# Train
model.fit(X_train, y_train)

# Prediction
y_pred = model.predict(X_test)

# Evaluation
k_accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", k_accuracy)

print("-----------------------------------------")

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("-----------------------------------------")

print("Classification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 0.8005
-----------------------------------------
Confusion Matrix:
[[1544   77]
 [ 322   57]]
-----------------------------------------
Classification Report:
              precision    recall  f1-score   support

           0       0.83      0.95      0.89      1621
           1       0.43      0.15      0.22       379

    accuracy                           0.80      2000
   macro avg       0.63      0.55      0.55      2000
weighted avg       0.75      0.80      0.76      2000



0 = No Heart Disease
1 = Heart Disease

Accuracy = 80.05%

1544 — Correctly identified healthy people

1,544 people did NOT have heart disease, and your model correctly predicted No Heart Disease.


77 — False alarms

77 people did NOT have heart disease, but your model predicted Heart Disease.

These are false positives.

322 — Missed heart disease cases

This is the important problem.

322 people actually HAD heart disease, but your model predicted No Heart Disease.

These are false negatives.

57 — Correctly detected heart disease

57 people actually had heart disease, and your model correctly predicted Heart Disease.

# HistGradientBoostingClassifier using OneHot Encoding

In [ ]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.ensemble import HistGradientBoostingClassifier
# Features
numeric_cols = [
    'Age',
    'Cholesterol',
    'Blood Pressure',
    'Heart Rate',
    'Exercise Hours',
    'Stress Level',
    'Blood Sugar'
]

categorical_cols = [
    'Gender',
    'Smoking',
    'Alcohol Intake',
    'Family History',
    'Obesity',
    'Exercise Induced Angina',
    'Chest Pain Type'
]

X = df[numeric_cols + categorical_cols]
y = df['Heart Disease']

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

# Preprocessing
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_cols),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
    ]
)

# Model pipeline
model = Pipeline([
    ('preprocessor', preprocessor),
    ('classifier', HistGradientBoostingClassifier(random_state=42))
])

# Train
model.fit(X_train, y_train)

# Prediction
y_pred = model.predict(X_test)

# Evaluation
h_accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", h_accuracy)

print("-----------------------------------------")

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("-----------------------------------------")

print("Classification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 0.817
-----------------------------------------
Confusion Matrix:
[[1555   66]
 [ 300   79]]
-----------------------------------------
Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.96      0.89      1621
           1       0.54      0.21      0.30       379

    accuracy                           0.82      2000
   macro avg       0.69      0.58      0.60      2000
weighted avg       0.78      0.82      0.78      2000



0 = No Heart Disease
1 = Heart Disease

Accuracy = 81.7%

1555 — Correctly identified healthy people

1,555 people did NOT have heart disease, and the model correctly predicted No Heart Disease.

66 — False alarms

66 people did NOT have heart disease, but the model predicted Heart Disease.

These are false positives.


300 — Missed heart disease cases

This is the most important problem.

300 people actually HAD heart disease, but the model predicted No Heart Disease.

These are false negatives. ⚠️


79 — Correctly detected heart disease

79 people actually had heart disease, and the model correctly predicted Heart Disease.

That's better than some of your previous models.

# Gradient Boosting using OneHot Encoding

In [ ]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.ensemble import GradientBoostingClassifier

# Features
numeric_cols = [
    'Age',
    'Cholesterol',
    'Blood Pressure',
    'Heart Rate',
    'Exercise Hours',
    'Stress Level',
    'Blood Sugar'
]

categorical_cols = [
    'Gender',
    'Smoking',
    'Alcohol Intake',
    'Family History',
    'Obesity',
    'Exercise Induced Angina',
    'Chest Pain Type'
]

X = df[numeric_cols + categorical_cols]
y = df['Heart Disease']

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

# Preprocessing
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_cols),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
    ]
)

# Model pipeline
model = Pipeline([
    ('preprocessor', preprocessor),
    ('classifier', GradientBoostingClassifier(random_state=42))
])

# Train
model.fit(X_train, y_train)

# Prediction
y_pred = model.predict(X_test)

# Evaluation
g_accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", g_accuracy)

print("-----------------------------------------")

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("-----------------------------------------")

print("Classification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 0.816
-----------------------------------------
Confusion Matrix:
[[1574   47]
 [ 321   58]]
-----------------------------------------
Classification Report:
              precision    recall  f1-score   support

           0       0.83      0.97      0.90      1621
           1       0.55      0.15      0.24       379

    accuracy                           0.82      2000
   macro avg       0.69      0.56      0.57      2000
weighted avg       0.78      0.82      0.77      2000



0 = No Heart Disease
1 = Heart Disease

Accuracy = 81.6%

1574 — Correctly identified healthy people

1,574 people did NOT have heart disease, and the model correctly predicted No Heart Disease.

47 — False alarms

47 people did NOT have heart disease, but the model predicted Heart Disease.

321 — Missed heart disease cases

This is the biggest problem.

321 people actually HAD heart disease, but the model predicted No Heart Disease.

These are false negatives. ⚠️

58 — Correctly detected heart disease

58 people actually had heart disease, and the model correctly predicted Heart Disease.

# Comparing all the results

In [ ]:
results=pd.DataFrame({
    'Model' : ['LogisticRegression','DecisionTreeClassifier','KNeighborsClassifier','HistGradientBoostingClassifier','GradientBoostingClassifier'],
    'Accuracy': [lo_accuracy,d_accuracy,k_accuracy,h_accuracy,g_accuracy]
})
print(results)

                            Model  Accuracy
0              LogisticRegression    0.8195
1          DecisionTreeClassifier    0.7130
2            KNeighborsClassifier    0.8005
3  HistGradientBoostingClassifier    0.8170
4      GradientBoostingClassifier    0.8160
